rules/copy_file_private.bzl

```diff
@@ -0,0 +1,121 @@
+# Copyright 2019 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Implementation of copy_file and copy_xfile rules.
+
+These rules copy a file to another location using Bash (on Linux/macOS) or
+cmd.exe (on Windows). 'copy_xfile' marks the resulting file executable,
+'copy_file' does not.
+"""
+
+def _common_impl(ctx, is_executable):
+    if ctx.attr.is_windows:
+        # Most Windows binaries built with MSVC use a certain argument quoting
+        # scheme. Bazel uses that scheme too to quote arguments. However,
+        # cmd.exe uses different semantics, so Bazel's quoting is wrong here.
+        # To fix that we write the command to a .bat file so no command line
+        # quoting or escaping is required.
+        bat = ctx.actions.declare_file(ctx.label.name + "-cmd.bat")
+        ctx.actions.write(
+            output = bat,
+            content = "@copy /Y \"%s\" \"%s\" >NUL" % (
+                ctx.file.src.path.replace("/", "\\"),
+                ctx.outputs.out.path.replace("/", "\\"),
+            ),
+            is_executable = True,
+        )
+        ctx.actions.run(
+            inputs = [ctx.file.src, bat],
+            outputs = [ctx.outputs.out],
+            executable = "cmd.exe",
+            arguments = ["/C", bat.path.replace("/", "\\")],
+            mnemonic = "StarlarkCopyFile",
+            progress_message = "Copying files",
+            use_default_shell_env = True,
+        )
+    else:
+        ctx.actions.run_shell(
+            inputs = [ctx.file.src],
+            outputs = [ctx.outputs.out],
+            command = "cp -f \"$1\" \"$2\"",
+            arguments = [ctx.file.src.path, ctx.outputs.out.path],
+            mnemonic = "StarlarkCopyFile",
```

Comment:
I don't think the word Starlark is useful here

Reply:
Removed.